### PR TITLE
Fix Journey Planner Back button behaviour

### PR DESCRIPTION
**Pull request by jezhiggins**
TapToRoute was rebuilding the undo stack on every restore, but should only do so when journey planning is still in progress. Fixed. 
Partly planned journeys could be lost if the app was paused and resumed. Also fixed.

Addressed issue #375

### CODE REVIEW

File: libraries/cyclestreets-view/src/main/java/net/cyclestreets/routing/Route.kt [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+    }
+
+    private fun setRouteLoaded() {
+        prefs()
+                .edit()
+                .putBoolean(routeLoadedPref, true)
+                .remove(waypointsInProgressPref)
+                .apply()
+    }
+    private fun clearRouteLoaded() {
```
Comment (oliverlockwood) [inline libraries/cyclestreets-view/src/main/java/net/cyclestreets/routing/Route.kt:180]: I find code generally easier to read with a blank line between consecutive methods.

File: libraries/cyclestreets-view/src/main/java/net/cyclestreets/routing/Route.kt [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+    }
+
+    private fun stashWaypoints() {
+        val stash = RouteDatabase.serializeWaypoints(waypoints_)
+        prefs()
+                .edit()
+                .putString(waypointsInProgressPref, stash)
+                .apply()
+    }
+    private fun restoreWaypoints() {
```
Comment (oliverlockwood) [inline libraries/cyclestreets-view/src/main/java/net/cyclestreets/routing/Route.kt:194]: as above - blank line between methods?

File: libraries/cyclestreets-view/src/main/java/net/cyclestreets/routing/Route.kt [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+    }
+
+    interface Listener {
+        fun onNewJourney(journey: Journey, waypoints: Waypoints)
+        fun onResetJourney()
+    }
+
+    private class Listeners {
+        private val listeners_: MutableList<Listener> = ArrayList()
+        fun register(listener: Listener) {
```
Comment (oliverlockwood) [inline libraries/cyclestreets-view/src/main/java/net/cyclestreets/routing/Route.kt:217]: I'd probably put a blank line before the method

File: libraries/cyclestreets-view/src/main/java/net/cyclestreets/routing/Route.kt [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+            if (!doRegister(listener)) return
+            if (journey() != NULL_JOURNEY || waypoints() != NULL_WAYPOINTS) listener.onNewJourney(journey(), waypoints()) else listener.onResetJourney()
```
Comment (oliverlockwood) [inline libraries/cyclestreets-view/src/main/java/net/cyclestreets/routing/Route.kt:219]: can we please break this line up like:
```
if (test || test)
  doThing()
else
  doOtherThing()
```

File: libraries/cyclestreets-view/src/main/java/net/cyclestreets/routing/Route.kt [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+            listeners_.add(listener)
+            return true
+        }
+
+        fun unregister(listener: Listener) {
+            listeners_.remove(listener)
+        }
+
+        fun onNewJourney(journey: Journey, waypoints: Waypoints) {
+            for (l in listeners_) l.onNewJourney(journey, waypoints)
```
Comment (oliverlockwood) [inline libraries/cyclestreets-view/src/main/java/net/cyclestreets/routing/Route.kt:237]: Could this be  like`listeners_.forEach(...)`?

File: libraries/cyclestreets-view/src/main/java/net/cyclestreets/routing/Route.kt [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+        }
+
+        fun onReset() {
+            for (l in listeners_) l.onResetJourney()
```
Comment (oliverlockwood) [inline libraries/cyclestreets-view/src/main/java/net/cyclestreets/routing/Route.kt:241]: listeners_.forEach(...)?

File: libraries/cyclestreets-view/src/main/java/net/cyclestreets/routing/Route.kt (hunk 2 above)
Comment (jezhiggins) [reply libraries/cyclestreets-view/src/main/java/net/cyclestreets/routing/Route.kt:194]: I had both these as single line methods, then jiggled them about and wasn't paying proper attention.

File: libraries/cyclestreets-view/src/main/java/net/cyclestreets/routing/Route.kt (hunk 4 above)
Comment (jezhiggins) [reply libraries/cyclestreets-view/src/main/java/net/cyclestreets/routing/Route.kt:219]: Certainly can

**Comment by oliverlockwood** [comment]
Looks good to me, and hacking about on the emulator I haven't been able to prove something is massively broken 😄

**Pull request merged**